File: framework/languages/dotnet/e2e/DiscoveryRegistryHa/Client/Scenarios/DrA2ClusterBridgeScenario.cs
using DiscoveryRegistryHa.Client.Support;
using DiscoveryRegistryHa.Shared;
using Zlink.HttpClient;

namespace DiscoveryRegistryHa.Client.Scenarios;

// DR-A2 verifies a reg-2-only consumer reaches a provider advertised through reg-1.
internal static class DrA2ClusterBridgeScenario
{
    public static async Task RunAsync(ClientOptions options)
    {
        using var reg2 = ZLinkHttpClient.Create(options.Reg2Url)
            .Json()
            .Timeout(TimeSpan.FromSeconds(10))
            .Build();
        using var consumer = ZLinkHttpClient.Create(options.ConsumerUrl)
            .Json()
            .Timeout(TimeSpan.FromSeconds(10))
            .Build();
        using var providerA = ZLinkHttpClient.Create(options.ProviderAUrl)
            .Json()
            .Timeout(TimeSpan.FromSeconds(10))
            .Build();

        await reg2.Post("/registry/members/wait")
            .Body(new MemberEndpointWaitRequest(options.ApiAEndpoint))
            .SubmitRawAsync();

        var marker = $"dr-a2-{Guid.NewGuid():N}";
        var reply = (await consumer.Post("/profile/request")
            .Body(new ProfileRequest("dr-a2", marker))
            .SubmitAsync<ProfileReply>()).Body;
        ScenarioAssert.That(reply.Value == "profile:dr-a2", "DR-A2 request failed.");
        ScenarioAssert.That(reply.ProviderRid == "api-a", "DR-A2 request should route to api-a.");
        ScenarioAssert.That(reply.Marker == marker, "DR-A2 marker mismatch.");

        var evidence = (await providerA.Post("/evidence/wait")
            .Body(new EvidenceWaitRequest(marker))
            .SubmitAsync<string[]>()).Body;
        ScenarioAssert.That(
            evidence.Any(line => line.Contains(marker, StringComparison.Ordinal)
                                 && line.Contains("rid=api-a", StringComparison.Ordinal)),
            "DR-A2 api-a evidence was not recorded.");

        Console.WriteLine("scenario DR-A2 passed");
    }
}

File: framework/languages/node/e2e/DiscoveryRegistryHa/Client/Scenarios/DrA2ClusterBridgeScenario.ts
import type { ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { getJson, postJson } from '../Support/http-client';
import { ensure } from '../Support/scenario-assert';
import { waitForMemberPeer } from '../Support/discovery-scenario-support';

export async function runDrA2(options: ClientOptions): Promise<void> {
  ensure(options.registry2Url !== undefined, 'DR-A2 requires registry-2-url.');

  const members = await waitForMemberPeer(options.registry2Url, 'api-a');
  ensure(members.some((entry) => String(entry.routingId) === 'api-a'), 'DR-A2 reg-2 member peer did not include api-a.');

  const reply = await postJson<ProfileRes>(options.consumerUrl, '/profile/request', { value: 'dr-a2' });
  ensure(reply.value === 'profile:dr-a2', 'DR-A2 reply value mismatch.');
  ensure(reply.providerRid === 'api-a', 'DR-A2 expected reg-1 provider api-a through reg-2 discovery.');

  const providerEvidence = await getJson<string[]>(options.providerAUrl, '/evidence');
  ensure(providerEvidence.some((entry) => entry.includes('value=dr-a2')), 'DR-A2 provider evidence missing.');
  console.log('scenario DR-A2 passed');
}
